src/nvim/api/buffer.c

```diff
@@ -161,64 +161,44 @@ Boolean nvim_buf_attach(uint64_t channel_id, Buffer buffer, Boolean send_buffer,
     return false;
   }
 
-  bool is_lua = (channel_id == LUA_INTERNAL_CALL);
   BufUpdateCallbacks cb = BUF_UPDATE_CALLBACKS_INIT;
-  struct {
-    const char *name;
-    LuaRef *dest;
-  } cbs[] = {
-    { "on_lines", &cb.on_lines },
-    { "on_bytes", &cb.on_bytes },
-    { "on_changedtick", &cb.on_changedtick },
-    { "on_detach", &cb.on_detach },
-    { "on_reload", &cb.on_reload },
-    { NULL, NULL },
-  };
-
-  for (size_t i = 0; i < opts.size; i++) {
-    String k = opts.items[i].key;
-    Object *v = &opts.items[i].value;
-    bool key_used = false;
-    if (is_lua) {
-      for (size_t j = 0; cbs[j].name; j++) {
-        if (strequal(cbs[j].name, k.data)) {
-          VALIDATE_T(cbs[j].name, kObjectTypeLuaRef, v->type, {
-            goto error;
-          });
-          *(cbs[j].dest) = v->data.luaref;
-          v->data.luaref = LUA_NOREF;
-          key_used = true;
-          break;
-        }
-      }
 
-      if (key_used) {
-        continue;
-      } else if (strequal("utf_sizes", k.data)) {
-        VALIDATE_T("utf_sizes", kObjectTypeBoolean, v->type, {
-          goto error;
-        });
-        cb.utf_sizes = v->data.boolean;
-        key_used = true;
-      } else if (strequal("preview", k.data)) {
-        VALIDATE_T("preview", kObjectTypeBoolean, v->type, {
-          goto error;
-        });
-        cb.preview = v->data.boolean;
-        key_used = true;
-      }
+  if (channel_id == LUA_INTERNAL_CALL) {
+    if (HAS_KEY(opts, buf_attach, on_lines)) {
+      cb.on_lines = opts->on_lines;
+      opts->on_lines = LUA_NOREF;
     }
 
-    VALIDATE_S(key_used, "'opts' key", k.data, {
-      goto error;
-    });
+    if (HAS_KEY(opts, buf_attach, on_bytes)) {
+      cb.on_bytes = opts->on_bytes;
+      opts->on_bytes = LUA_NOREF;
+    }
+
+    if (HAS_KEY(opts, buf_attach, on_changedtick)) {
+      cb.on_changedtick = opts->on_changedtick;
+      opts->on_changedtick = LUA_NOREF;
+    }
+
+    if (HAS_KEY(opts, buf_attach, on_detach)) {
+      cb.on_detach = opts->on_detach;
+      opts->on_detach = LUA_NOREF;
+    }
+
+    if (HAS_KEY(opts, buf_attach, on_reload)) {
+      cb.on_reload = opts->on_reload;
+      opts->on_reload = LUA_NOREF;
+    }
+
+    if (HAS_KEY(opts, buf_attach, utf_sizes)) {
+      cb.utf_sizes = opts->utf_sizes;
+    }
+
+    if (HAS_KEY(opts, buf_attach, preview)) {
+      cb.preview = opts->preview;
+    }
```

Comment:
these both default to `false` so the assignments can be done directly without HAS_KEY() check

Reply:
Fixed in 67f1fa8940ed14bdcf221383b6033c9391b93d77.